gRPC API & internals redesign

KeyboardNerd opened the pull request:
This is a large Pull Request that will contain a number of chunks. For easier code review, the Clair V3 PR will contains these commits separately.
1. API changes
2. datastore interface changes
3. mock datastore, extensions update, worker, updater, notifier changes with tests
4. datastore pgsql implementation with tests
5. potential vendor clean up

KeyboardNerd committed:
API: drop v1 api, changed v2 api for Clair v3.

jzelinskie commented on api/v2/clairpb/clair.proto:
```diff
@@ -26,57 +27,63 @@ message Vulnerability {
 	string link = 4;
 	string severity = 5;
 	string metadata = 6;
+	// fixed_by only exists when querying features
```
Can we treat comments like Go ones? Full sentences 80 char lines.

jzelinskie commented on api/v2/clairpb/clair.proto:
```diff
@@ -26,57 +27,63 @@ message Vulnerability {
 
-message Feature {
+message ClairStatus {
+	// all listers and detectors implemented in this clair version
```
Is this what Clair is built with or configured with?

KeyboardNerd replied:
all built-in listers/detectors. The listers and detectors are not configurable now, should I add that to this PR?

jzelinskie replied:
Nope, we can do that later, just as long as nothing here prevents us from doing so.

jzelinskie commented on api/v2/clairpb/clair.proto:
```diff
@@ -26,57 +27,63 @@ message Vulnerability {
+
+message PagedVulnerableAncestries {
 	string this_token = 1;
 	string next_token = 2;
```
maybe rename these 'current_page' and 'next_page'

jzelinskie commented on api/v2/rpc.go:
```diff
@@ -164,93 +193,23 @@ func (s *NotificationServer) DeleteNotification(ctx context.Context, req *pb.Del
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
```
any reason you don't defer the rollback here and instead explicitly call it for each branch below?

KeyboardNerd replied:
I don't think Rollback() should ever be called after Commit() on success.

jzelinskie replied:
Rollback() from database/sql actually is safe to do this and is considered an idiomatic pattern to do so. I think it makes sense both ways, though.

jzelinskie commented on database/models.go:
```diff
@@ -17,52 +17,108 @@ package database
 
-	// For output purposes. Only make sense when the feature version is in the context of an image.
-	AddedBy Layer
+// AffectedFeature is used determine whether a namespaced feature is affected.
```
AffectedFeature is used to determine whether a namespaced feature is affected by a Vulnerability.

KeyboardNerd committed:
Datastore: updated for Clair V3, decoupled interfaces and models

KeyboardNerd committed:
Clair Logic, Extensions: updated mock tests, extensions, basic logic

Main Clair logic is changed in worker, updater, notifier for better adapting
ancestry schema. Extensions are updated with the new model and feature lister
 and namespace detector drivers are able to specify the specific listers and
detectors used to process layer's content. InRange and GetFixedIn interfaces
are added to Version format for adapting ranged affected features and next
available fixed in in the future. Tests for worker, updater and extensions
are fixed.

KeyboardNerd committed:
database: postgres implementation with tests.

KeyboardNerd commented:
I'm using my analyze_local_images fork's py tools to test the APIs, which seems to work properly. I'll do more concurrency tests and the web hook tests.

Besides the implementation of new schema and analyzing functions, I also added the knob to control feature listers and namespace detectors to scan the layers and updaters to fetch vuln src.

jzelinskie commented on api/v2/clairpb/convert.go:
```diff
@@ -18,36 +18,55 @@ import (
+		return nil, err
+	}
+
+	// encrypt page numbers
```
this isn't a particularly useful comment

jzelinskie commented on api/v2/clairpb/convert.go:
```diff
@@ -18,36 +18,55 @@ import (
+	}
+
+	// encrypt page numbers
+	nextToken, err := dbVuln.Next.Encrypt(paginationKey)
```
What happens when there is no next page? Is there a special constant for that or does this function need to handle `nil`?

jzelinskie commented on api/v2/clairpb/convert.go:
```diff
@@ -59,35 +78,27 @@ func NotificationFromDatabaseModel(dbNotification database.VulnerabilityNotifica
-		},
-	}, nil
-}
+	pbOldVuln, err := PagedVulnerableAncestriesFromDatabaseModel(dbOldVuln, paginationKey)
```
What happens if there is no old vulnerability? Doesn't this need to handle `nil`?

jzelinskie commented on api/v2/rpc.go:
```diff
@@ -46,211 +43,221 @@ type AncestryServer struct {
+		}
+
+		if !ok {
+			return nil, status.Error(codes.NotFound, "requested ancestry \""+ancestry.Name+" \" is not found")
```
raw string literals will make this easier to read

```go
return nil, status.Error(`requested ancestry "` + ancestry.Name + `" is not found`)
```

jzelinskie commented on api/v2/rpc.go:
```diff
@@ -46,211 +43,221 @@ type AncestryServer struct {
+				return nil, status.Error(codes.Internal, err.Error())
+			}
+
+			// ensure every feature is found
```
```go
// Ensure that every feature can be found.
```

I'd also move this comment to be above `if !fv.Valid` instead of the whole for loop.

jzelinskie commented on database/models.go:
```diff
@@ -17,52 +17,116 @@ package database
+type AffectedFeature struct {
+	Namespace   Namespace
+	FeatureName string
+	// known fixed in version for the vulnerability
```
full sentences for comments

jzelinskie commented on database/models.go:
```diff
@@ -71,17 +135,78 @@ type Vulnerability struct {
 	Metadata MetadataMap
+}
+
+// VulnerabilityWithAffected is an vulnerability with all known affected
```
**a** vulnerability

jzelinskie commented on database/models.go:
```diff
@@ -71,17 +135,78 @@ type Vulnerability struct {
+	Next    PageNumber
 }
 
+// Notification is a hook used to notify and give token to user to retrieve the
```
Let's rename this NotificationHook.

NotificationHook is a message sent to another service to inform of a change to a Vulnerability or the Ancestries affected by a Vulnerability. It contains the name of a notification that should be read and marked as read via the API.

jzelinskie commented on database/models.go:
```diff
@@ -71,17 +135,78 @@ type Vulnerability struct {
+type PageNumber struct {
+	// StartID is an implementation detail for paginating by an ID required to
+	// be unique to every ancestry and always increasing.
+	StartID int
```
Ideally, this is just a string and we encrypted this before it ever leaves the interface, making it an opaque string to everyone but the implementation of the data store interface.

KeyboardNerd replied:
ok, so in that case the implementation should initialize an encrypted version of
 VulnerabilityNotificationFirstPage and VulnerabilityNotificationLastPage 
when it's registered. And the pagination key should be kept being used only inside pgsql.

jzelinskie commented on pkg/strutil/strutil.go:
```diff
@@ -0,0 +1,41 @@
+package strutil
```
copyright header

jzelinskie commented on pkg/strutil/strutil_test.go:
```diff
@@ -0,0 +1,20 @@
+package strutil
```
copyright header

jzelinskie commented on cmd/clair/main.go:
```diff
@@ -30,26 +30,30 @@ import (
 	_ "github.com/coreos/clair/ext/featurefmt/apk"
 	_ "github.com/coreos/clair/ext/featurefmt/dpkg"
 	_ "github.com/coreos/clair/ext/featurefmt/rpm"
+	"github.com/coreos/clair/ext/featurens"
```
import these with `strutil`, `stopper`, etc... leave these only for the side-effect imports that require `_`

jzelinskie commented on cmd/clair/main.go:
```diff
@@ -85,6 +89,43 @@ func stopCPUProfiling(f *os.File) {
 }
 
+func configClairVersion(config *Config) {
+	// prop to user of the current clair registered compontent
```
I'm not sure what you're trying to say here. Not sure if `prop` is the word you're looking for.

jzelinskie commented on config.example.yaml:
```diff
@@ -51,10 +51,29 @@ clair:
     certfile:
 
+  worker:
+    enableddetectors:
```
let's call these `namespace_detectors`

jzelinskie commented on config.example.yaml:
```diff
@@ -51,10 +51,29 @@ clair:
+      - alpine-release
+      - redhat-release
+
+    enabledlisters:
```
let's call this `feature_listers`

jzelinskie commented on ext/featurefmt/apk/apk_test.go:
```diff
@@ -19,58 +19,65 @@ import (
+		{
+			Name:    "musl",
+			Version: "1.1.14-r10",
+		},
```
Does gofmt prevent this?

```go
testFeatures := []database.Feature{
  {Name: "musl", Version: "1.1.14-r10"},
  {Name: "busybox", Version: "1.24.2-r9"},
}
```

jzelinskie commented on ext/featurefmt/driver.go:
```diff
@@ -117,10 +107,19 @@ func RequiredFilenames() (files []string) {
 	return
 }
 
+// ListListers returns a list of all listers' names
```
Let's avoid a tongue twister here:

>ListListers returns the names of all the registered feature listers.

jzelinskie commented on ext/featurens/driver.go:
```diff
@@ -106,6 +110,14 @@ func RequiredFilenames() (files []string) {
 	return
 }
 
+func ListDetectors() []string {
```
>ListDectectors returns the names of all registered namespace detectors.

jzelinskie commented on ext/versionfmt/rpm/parser.go:
```diff
@@ -121,6 +121,19 @@ func (p parser) Valid(str string) bool {
+	return cmp < 0, nil
+}
+
+// only for temporary!
```
o rly

jzelinskie commented on ext/vulnsrc/debian/debian.go:
```diff
@@ -62,19 +62,32 @@ func init() {
 		return resp, err
 	}
 
+	if err := tx.Rollback(); err != nil {
```
why is this rolling back always here?

KeyboardNerd replied:
Because the transaction won't mutate the database

jzelinskie commented on ext/vulnsrc/driver.go:
```diff
@@ -88,3 +87,11 @@ func Updaters() map[string]Updater {
 	return ret
 }
+
+func ListUpdaters() []string {
```
>ListUpdaters returns the names of registered vulnerability updaters.

jzelinskie commented on ext/vulnsrc/rhel/rhel.go:
```diff
@@ -90,11 +90,24 @@ func init() {
 		return resp, err
 	}
+
+	tx.Rollback()
```
why always rollback here?

jzelinskie commented on ext/vulnsrc/ubuntu/ubuntu.go:
```diff
@@ -98,12 +98,22 @@ func (u *updater) Update(datastore database.Datastore) (resp vulnsrc.UpdateRespo
 		return resp, err
 	}
 
+	tx.Rollback()
```
why always rollback here?

jzelinskie commented on updater.go:
```diff
@@ -349,39 +387,313 @@ func (lv *lockableVulnerability) appendFunc(metadataKey string, metadata interfa
+		log.WithError(err).Error()
+	}
+	defer tx.Rollback()
+	return tx.FindLock(updaterLockName)
```
this is missing a commit?

jzelinskie commented on updater.go:
```diff
@@ -349,39 +387,313 @@ func (lv *lockableVulnerability) appendFunc(metadataKey string, metadata interfa
+	if err != nil {
+		return err
+	}
+
```
why not defer rollback?

jzelinskie commented on updater.go:
```diff
@@ -349,39 +387,313 @@ func (lv *lockableVulnerability) appendFunc(metadataKey string, metadata interfa
+	if err != nil {
+		return nil, err
+	}
+
```
why not defer rollback?

jzelinskie commented on worker.go:
```diff
@@ -44,177 +43,458 @@ var (
 	}
 
-	log.WithFields(log.Fields{logLayerName: name, "path": cleanURL(path), "engine version": Version, "parent layer": parentName, "format": imageFormat}).Debug("processing layer")
+	// TODO: to fix, when there are duplicated layers
```
hmm

KeyboardNerd replied:
actually it's not true! the layers are deduplicated before the function call

jzelinskie commented on worker.go:
```diff
@@ -44,177 +43,458 @@ var (
 	if err != nil {
-		return
+		return err
 	}
```
defer rollback?

jzelinskie commented on worker.go:
```diff
@@ -44,177 +43,458 @@ var (
+	// version format -> feature ID -> feature
+	features := map[string]map[string]database.NamespacedFeature{}
+	for _, layer := range ancestryLayers {
+		// at start of the loop, namespaces and features always contain the previous
```
capital at the start of sentences, end with periods, lol grammar

jzelinskie commented on worker_test.go:
```diff
@@ -15,61 +15,324 @@
+	"github.com/coreos/clair/ext/featurefmt"
 	_ "github.com/coreos/clair/ext/featurefmt/dpkg"
+	_ "github.com/coreos/clair/ext/featurefmt/rpm"
+	"github.com/coreos/clair/ext/featurens"
```
pull these up to where they used to be

jzelinskie reviewed (changes requested):
oh boy, here we go

KeyboardNerd committed:
pgSQL: fixed invalidating vulnerability cache query.

jzelinskie commented:
I think we should rename the API package to just "api" or "v3" (because Clair v3) in order to avoid confusion.

After that I think this is safe to merge!

KeyboardNerd committed:
api: renamed V2 API to V3 API for consistency.
